server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -669,8 +598,97 @@ private void updateMergeState(Map<TableId,MergeStats> mergeStatsCache) {
+          deletionStartRow);
+      final KeyExtent firstTablet =
+          loadTabletMetadata(info.getExtent().tableId(), deletionStartRow, ColumnType.PREV_ROW)
+              .map(TabletMetadata::getExtent).orElse(null);
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Skip to the next highest
+      // tablet after this one if it exists
+      if (needsFencingForDeletion(info, firstTablet)) {
+        Key nextExtent = new Key(deletionStartRow).followingKey(PartialKey.ROW);
+        deletionStartRow = getHighTablet(new KeyExtent(info.getExtent().tableId(),
+            nextExtent.getRow(), info.getExtent().prevEndRow())).endRow();
+        Manager.log.trace("First tablet in deletion range updated to contain startRow {}",
+            deletionStartRow);
+      }
+    }
+    return deletionStartRow;
+  }
+
+  private Text getDeletionEndRow(MergeInfo info) {
+    Text deletionEndRow = info.getExtent().endRow();
+    if (deletionEndRow != null) {
+      Manager.log.trace("Finding tablet that contains the end row {} for deletion", deletionEndRow);
+      final KeyExtent last =
+          loadTabletMetadata(info.getExtent().tableId(), deletionEndRow, ColumnType.PREV_ROW)
+              .map(TabletMetadata::getExtent).orElse(null);
+      // If needsFencingForDeletion() is true then this tablet was previously fenced
+      // so we need to keep it and not delete the tablet. Find the previous tablet
+      // if it exists to stop deletion on
+      if (needsFencingForDeletion(info, last)) {
+        deletionEndRow = last.prevEndRow();
+        Manager.log.debug("Last tablet in deletion range updated to contain endRow {}",
+            deletionEndRow);
+      }
+    }
+    return deletionEndRow;
+  }
+
+  private boolean hasTabletsToDelete(MergeInfo info, Text deletionStartRow, Text deletionEndRow) {
+    // If the endRow of the deletion range is not null, but the actual deletionEndRow is null then
+    // this means
+    // the first tablet was fenced so there are no tablets to delete and we can return
+    if (info.getExtent().endRow() != null && deletionEndRow == null) {
+      return false;
+      // If the prevEndRow of the deletion range is not null, but the actual deletionStartRow is
+      // null then this means
+      // the last tablet was fenced so there are no tablets to delete and we can return
+    } else if (info.getExtent().prevEndRow() != null && deletionStartRow == null) {
+      return false;
+      // Make sure detected deletionEndRow is after deletionStartRow. If they are the same tablet or
+      // before then nothing to delete and we can return
+      // This would happen in some situations such as a deletion range fully contained in 1 tablet
+    } else if (deletionEndRow != null && deletionStartRow != null
+        && new Key(deletionEndRow).compareRow(deletionStartRow) <= 0) {
+      return false;
+    }
+
+    return true;
```

Comment:
Are the following the only two case where there are not tablet to delete?

 1. The case where delete range falls completely within a single tablet.
 2. The case where the delete range overlaps two contiguous tablets.  For example a table with splits `[c,m,s]` and a delete range of `(f,q]` then the tablet with end row `m` and `s` would each be fenced.  So those two tablets are contiguous with nothing between them.

For both cases above is the assumption that deletionStartRow and deletionEndRow are equal?  I think that is the case.


I am wondering if this code could do the following instead where that code handles nulls. But not sure about this, just thinking out loud.  

```suggestion
  private boolean hasTabletsToDelete(MergeInfo info, Text deletionStartRow, Text deletionEndRow) {
     return !Objects.equals(deletionStartRow, deletionEndRow);
  }
```

Also I do not understand the case in last else if it where it does ``new Key(deletionEndRow).compareRow(deletionStartRow) <= 0`, not sure about this less than or equals check and how it fits into the bigger picture.

Reply:
@keith-turner - I don't think your simplification entirely works (I tried it and it fails with errors), It's been a couple weeks but I know I added the checks here for the various edge cases I encountered and with comments about how, however it's possible that this could be re-worked so it is easier to understand and maybe a slight tweak to what you did will work.

I started working through this a bit again using the comments I left as a guide and I am going back through the tests to look at the different edge cases that caused these scenarios and I can report back what deletion range caused each one and why.  After that we can look at if there is a better way.

Reply:
@keith-turner - Ok I went through this method and I updated it in https://github.com/apache/accumulo/pull/3728/commits/f1887245213c65f1a415c60e0a39818fb53e37a3 to be a lot more clear and simpler. 

There are 4 conditions where we don't have any tablets to delete and only fencing is needed:

1. If the first and last tablets that overlap the deletion range are equal then the range is inside a single tablet that is fenced.
2. If the last tablet that overlaps the deletion range is the first tablet in the table then there is nothing before it to delete and that tablet is fenced.
3. If the first tablet that overlaps the deletion range is the last tablet in the table then there is nothing after it to delete and that tablet is fenced.
4. If the first and last tablet that overlap the deletion range are contiguous then both are fenced and nothing in between to delete.

Reply:
I pushed one more commit to rename things to be more clear: https://github.com/apache/accumulo/pull/3728/commits/0d744b9b2fa1a05a4904606629c1da78d64f80b0

Reply:
Those changes are really nice.

Reply:
@cshannon  curious if the test cover all the cases you mentioned above.

Reply:
> @cshannon curious if the test cover all the cases you mentioned above.

Yep, the DeleteRowsIT covers all the cases, I went through it when I made my latest changes to make sure everything looked good. There's a test that eliminates some rows in only one tablet, one that eliminates some rows in the first tablet, another one that eliminates some rows in the last tablet and a test that causes two tablets to be contiguous and fenced with nothing in the middle. I can double check though to see if we need more coverage. I'm planning follow on PRs anyways for more test coverage (i really want to exercise and make sure all edge cases are good)